mysql-test/suite/binlog/r/binlog_recover_async_rollback_trx.result

```diff
@@ -0,0 +1,64 @@
+CREATE TABLE t1 (id int primary key, data int) ENGINE = InnoDB;
+INSERT INTO t1 VALUES (0, 1);
+#
+# 1. Check DML in prepared state can rollback correctly.
+#
+connect con1, localhost, root,,;
+SET debug_sync = "ha_commit_trans_after_prepare SIGNAL prepared1 WAIT_FOR continue";
+INSERT INTO t1 VALUES(1, 1);;
+connect con2, localhost, root,,;
+SET debug_sync = "now WAIT_FOR prepared1";
+SET debug_sync = "ha_commit_trans_after_prepare SIGNAL prepared2 WAIT_FOR continue";
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+connection default;
+SET debug_sync = "now WAIT_FOR prepared2";
+# Kill the server
+disconnect con1;
+disconnect con2;
+# restart: --debug=d,debug_innodb_rollback_thread --log-error=MYSQLTEST_VARDIR/tmp/log.err
+include/assert_grep.inc [Server reported: recovery_rollback_thread]
+INSERT INTO t1 VALUES(1, 1);
+# Expected (0, 1), (1, 1)
+SELECT * FROM t1;
+id	data
+0	1
+1	1
+# Expect (0,2), (1,1)
+SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
+SELECT * FROM t1;
+id	data
+0	2
+1	1
+#
+# 2. Check DML can rollback automatically after second crash.
+#
+# Kill the server
+# restart
+include/assert_grep.inc [Server reported: Rolled back recovered transaction]
+include/assert_grep.inc [Server NOT report: Found 1 prepared transaction]
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+# expected (0, 2), (1, 1)
+SELECT * FROM t1;
+id	data
+0	2
+1	1
+# 3. Test that innodb shutdown as expected if any error happens before
+#    normal rollback task is started. In the situation, rollback task
+#    should be started at preshutdown accordingly to rollback or
+#    deregister all recovered active transactions.
+BEGIN;
+INSERT INTO t1 SELECT id + power(2, 2), 1 FROM t1;
+INSERT INTO t1 SELECT id + power(2, 3), 1 FROM t1;
+INSERT INTO t1 SELECT id + power(2, 4), 1 FROM t1;
+INSERT INTO t1 SELECT id + power(2, 5), 1 FROM t1;
+INSERT INTO t1 SELECT id + power(2, 6), 1 FROM t1;
+INSERT INTO t1 SELECT id + power(2, 7), 1 FROM t1;
+COMMIT;
```

Comment:
For the `SELECT` part of this, explicit record locks will unnecessarily be allocated until [MDEV-24813](https://jira.mariadb.org/browse/MDEV-24813) has been fixed. See 516f7c111b0cc3e9ad193c5a4e8c9171a081a362 and e98a5e166c16f24fef2d4f356b0d82559a835ee4 how to use the sequence engine to work around this deficiency.

Reply:
fine, use sequence instead.